### PR TITLE
feat(scheduler): Add topology aware scenario filter

### PR DESCRIPTION
**Pull request by davidLif**
## Description

Add basic topology aware scenario filter - for each topology constraint with "required", check if a domain exists on this level with enough free gpus. Like the cluster-wide idle gpus filter, check that the capacty comparisons we look at all the request on the same level and make sure all of them can be satisfied together

## Related Issues

Fixes #

## Checklist

> **Note:** Ensure your PR title follows the [Conventional Commits format](https://github.com/NVIDIA/KAI-Scheduler/blob/main/CONTRIBUTING.md#pr-title-guidelines) (e.g., `feat(scheduler): add new feature`)

- [x] Self-reviewed
- [x] Added/updated tests (if needed)
- [ ] Updated documentation (if needed)

## Summary by CodeRabbit

* **New Features**
  * Added topology-aware GPU idle-capacity filtering so scheduling respects topology domains and constraints.

* **Refactor**
  * Streamlined GPU allocation and victim-task processing for more reliable, efficient capacity accounting and to avoid double-counting freed GPUs.

* **Tests**
  * Added comprehensive tests covering topology constraints, victim handling, multi-level domains, and allocation edge cases.

### CODE REVIEW

**Comment by davidLif** [comment]
@coderabbitai review

File: pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus/common.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+// descending. Returns true if all non-zero requirements can be matched.
+func greedyMatchRequirements[K comparable](
+	requirements []float64,
+	holders []K,
+	capacity func(K) float64,
+) bool {
+	virtuallyAllocated := make(map[K]float64, len(holders))
+	for _, required := range requirements {
+		if required == 0 {
+			continue
```
Comment (gshaibi) [inline pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus/common.go]: Assuming the requirements are sorted, can we return true here?

File: pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus/topology_aware_idle_gpus_test.go [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+		"node-3": {topologyLabelKey: "rack-3"},
+	}
+	nodes := createTestNodes(ctrl, map[string]int{
+		"node-1": 4,
+		"node-2": 5,
+		"node-3": 6,
+	}, topologyLabels)
+
+	rootSubGroupSet := subgroup_info.NewSubGroupSet(subgroup_info.RootSubGroupSetName, nil)
+	rootSubGroupSet.AddSubGroup(newConstrainedSubGroup(podgroup_info.DefaultSubGroup, topology, requiredLevel, 2))
```
Comment (gshaibi) [inline pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus/topology_aware_idle_gpus_test.go]: Should it be like so?
```suggestion
	rootSubGroupSet.AddSubGroup(newConstrainedSubGroup(podgroup_info.DefaultSubGroup, topology, topologyLabelKey, 2))
```

File: pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus/common.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+			if totalCapacity < required {
+				break
+			}
+			available := totalCapacity - virtuallyAllocated[holder]
+			if available >= required {
+				virtuallyAllocated[holder] += required
+				matched = true
+				break
+			}
+		}
```
Comment (gshaibi) [inline pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus/common.go:58]: nit - will this work? (without the matched variable)
```suggestion
		for _, holder := range holders {
			available := capacity(holder) - virtuallyAllocated[holder]
			if available < required {
                return false
			}
            virtuallyAllocated[holder] += required
		}
```

File: pkg/scheduler/actions/common/solvers/pod_scenario_builder.go [hunk 4]
```diff
@@ -7,6 +7,7 @@ import (
 	"golang.org/x/exp/slices"
 
 	"github.com/NVIDIA/KAI-scheduler/pkg/scheduler/actions/common/solvers/accumulated_scenario_filters"
+	idle_gpus_filer "github.com/NVIDIA/KAI-scheduler/pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus"
```
[the remark this answers is not in the document]
Comment (gshaibi) [reply pkg/scheduler/actions/common/solvers/pod_scenario_builder.go]: typo - `idle_gpus_filter`

[the remark this answers is not in the document]
Comment (davidLif) [reply pkg/scheduler/actions/common/solvers/pod_scenario_builder.go]: Fixed

**Comment by davidLif** [comment]
@coderabbitai full review

File: pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus/topology_aware_idle_gpus_test.go [hunk 5]
```diff
@@ -0,0 +1,579 @@
...
+		[]*pod_info.PodInfo{victimTask},
+		nil,
+	)
+
+	filter := NewTopologyAwareIdleGpusFilter(testScenario, nodes)
+	if filter == nil {
+		t.Fatal("Expected non-nil filter")
+	}
+	runFilterCheck(t, filter, testScenario, true, "rack-1 will have 12 GPUs after evicting victim")
+}
```
[the remark this answers is not in the document]
Comment (davidLif) [reply pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus/topology_aware_idle_gpus_test.go:209]: Will move victim to node2

File: pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus/topology_aware_idle_gpus.go [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+		domainsByConstraint[cgKey] = append(domainsByConstraint[cgKey], key)
+	}
+	for _, domains := range domainsByConstraint {
+		sort.Slice(domains, func(i, j int) bool {
+			return domainCapacity[domains[i]] > domainCapacity[domains[j]]
+		})
+	}
+
+	return domainCapacity, nodeToTopologyDomains, domainsByConstraint
+}
```
[the remark this answers is not in the document]
Comment (davidLif) [reply pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus/topology_aware_idle_gpus.go:248]: No relevnat. Nill nodes are filtered long before from the session

File: pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus/common.go (hunk 3 above)
Comment (davidLif) [reply pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus/common.go:58]: No. The "return might be too early, because the "virtuallyAllocated" for holder i might be much beggier then i+1, so that `capacity(holder[i]) - virtuallyAllocated[holder[i]] < required ` but `capacity(holder[i+1]) - virtuallyAllocated[holder[i+1]] > required `. This will not hold true when just loooking at the capacity, so `totalCapacity < required` allows us to exit early.

File: pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus/common.go (hunk 1 above)
Comment (davidLif) [reply pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus/common.go]: This does look true. Will add a fix.

File: pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus/topology_aware_idle_gpus_test.go (hunk 2 above)
Comment (davidLif) [reply pkg/scheduler/actions/common/solvers/accumulated_scenario_filters/idle_gpus/topology_aware_idle_gpus_test.go]: Fixed
